src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2312 @@
+        addLog(LOG_LEVEL_INFO, log);
+      }
+      # endif // ifdef P104_DEBUG_DEV
+
+      zones.clear();
+      zonesInitialized = false;
+      numDevices       = 0;
+
+      int16_t  offset2;
+      uint16_t prev2 = 0;
+      String   tmp, fld;
+      int tmp_int;
+      offset2 = buffer.indexOf(P104_ZONE_SEP);
+
+      while (offset2 > -1) {
+        tmp.reserve(offset2 - prev2);
+        tmp = buffer.substring(prev2, offset2);
+
+        zones.push_back(P104_zone_struct(zoneIndex + 1));
+
+        tmp_int = 0;
+
+        // WARNING: Order of values should match the numeric order of P104_OFFSET_* values
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_SIZE, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].size = tmp_int;
+        }
+
+        zones[zoneIndex].text = parseStringKeepCase(tmp, 1 + P104_OFFSET_TEXT, P104_FIELD_SEP);
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_ALIGNMENT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].alignment = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_ANIM_IN, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].animationIn = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_SPEED, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].speed = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_ANIM_OUT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].animationOut = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_PAUSE, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].pause = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_FONT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].font = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_CONTENT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].content = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_LAYOUT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].layout = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_SPEC_EFFECT, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].specialEffect = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_OFFSET, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].offset = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_BRIGHTNESS, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].brightness = tmp_int;
+        }
+
+        if (validIntFromString(parseString(tmp, 1 + P104_OFFSET_REPEATDELAY, P104_FIELD_SEP), tmp_int)) {
+          zones[zoneIndex].repeatDelay = tmp_int;
+        }
+
+        delay(0);
+        prev2       = offset2 + 1;
+        offset2     = buffer.indexOf(P104_ZONE_SEP, prev2);
```

Comment:
This `buffer` looks like it is being used as a `String[]`
So why not use it like that and use the `SaveStringArray` function?
Then you don't need to allocate large blocks to make sure it all fits and it is still rather dynamic in how it is stored.
Then you also don't need to create `tmp` strings to define a new zone.